src/main/java/org/springframework/data/couchbase/core/RxJavaCouchbaseTemplate.java

```diff
@@ -102,6 +113,74 @@
                 .flatMap(object -> remove(object, persistTo, replicateTo));
     }
 
+    @Override
+    public <T> Observable<T> cas(String id, Class<T> entityClass, Function<Optional<T>, CasResult<T>> updateFunction) {
+        return cas(id, entityClass, PersistTo.NONE, ReplicateTo.NONE, updateFunction);
+    }
+
+    @Override
+    public <T> Observable<T> cas(String id, Class<T> entityClass, PersistTo persistTo, ReplicateTo replicateTo, Function<Optional<T>, CasResult<T>> updateFunction) {
+        return asyncCas(id, entityClass, persistTo, replicateTo, old -> Observable.just(updateFunction.apply(old)));
+    }
+
+    @Override
+    public <T> Observable<T> asyncCas(String id, Class<T> entityClass, Function<Optional<T>, Observable<CasResult<T>>> updateFunction) {
+        return asyncCas(id, entityClass, PersistTo.NONE, ReplicateTo.NONE, updateFunction);
+    }
+
+    @Override
+    public <T> Observable<T> asyncCas(String id, Class<T> entityClass, PersistTo persistTo, ReplicateTo replicateTo, Function<Optional<T>, Observable<CasResult<T>>> updateFunction) {
+        return client.get(id, RawJsonDocument.class)
+                .flatMap(doc -> update(doc, entityClass, persistTo, replicateTo, updateFunction))
+                .switchIfEmpty(insert(entityClass, persistTo, replicateTo, updateFunction))
+                .retryWhen(casRetryFunction())
+                .onErrorResumeNext(e -> Observable.error(e.getCause()))
+                .doOnError(e -> TemplateUtils.translateError(e));
+    }
+
+    private <T> Observable<T> insert(Class<T> entityClass, PersistTo persistTo, ReplicateTo replicateTo, Function<Optional<T>, Observable<CasResult<T>>> updateFunction) {
+        return Observable.fromCallable(Optional::<T>empty)
+                .compose(doModification(entityClass, updateFunction, newDoc -> client.insert(newDoc, persistTo, replicateTo)));
+    }
+
+    private <T> Observable<T> update(RawJsonDocument oldDoc, Class<T> entityClass, PersistTo persistTo, ReplicateTo replicateTo, Function<Optional<T>, Observable<CasResult<T>>> updateFunction) {
+        T entity = mapToEntity(oldDoc.id(), oldDoc, entityClass);
+        return Observable.just(entity)
+                .map(Optional::ofNullable)
+                .compose(doModification(entityClass, updateFunction, newDoc -> client.replace(newDoc, persistTo, replicateTo)))
+                .switchIfEmpty(Observable.just(entity));
+    }
+
+    private <T> Observable.Transformer<Optional<T>, T> doModification(
+            Class<T> entityClass,
+            Function<Optional<T>, Observable<CasResult<T>>> updateFunction,
+            Function<RawJsonDocument, Observable<RawJsonDocument>> modificationFunction) {
+        return observable -> observable
+                .flatMap(updateFunction::apply)
+                .filter(this::shouldUpdate)
```

Comment:
I think bubbling up the CASMismatchException, .. other exceptions in the casRetryFunction should be good enough. Spring retry can be used to do auto-retries, we don't usually provide this internally with the sdk itself as it is easily handled on the application end.